## Sprint 2 Data - Omisha Mondal DS4003

## 1. Data and Data Provenance:

The dataset I am exploring comes from Yahoo Finance. It is industry stock market data. I found it on Kaggle at the following link: https://www.kaggle.com/datasets/belayethossainds/yahoo-finance-industries-dataset.

The reason I chose this dataset is because it will allow me to create significant visualizations relating to my initial user stories. It will give them more perspective on stocks that they may be unfamiliar with, while also educating them on different companies across industries. Each stock also has information about its market capitalizations, which can allow potential investors to make informed decisions about whether they would like to invest in the company or not. Also, important columns like change shows the daily change, which would be relevent for users who are interested in investing for long periods of time. The prices of the stocks can allow for potential investors to see whether they can diversify their portfolio. PE Ratios can also help the users to understand the market conditions behind the stocks. Overall, this dataset is ideal because it had several key columns that provides useful information which can allow me to create an interactive dashboard with stock trends and recommendations.

Data Provenance: The data originates from Yahoo Finance, specifically the Industries category. https://www.yahoo.com/author/yahoo-finance/. The collection methodology is Selenium Automation scrapping method. The collaborator is Belayet Hossain, who is also the owner. He is a web scraper. The coverage start date is on 5/13/2023, and the coverage end date is also 5/13/2023.
DOI Citation: Belayet Hossain. (2023). Yahoo Finance - Industries - Dataset [Data set]. Kaggle. https://doi.org/10.34740/KAGGLE/DSV/5678079

In [17]:
# Import dependencies
import pandas as pd
import plotly.express as px

# Dataset
data = pd.read_csv("yahoo Finanace_industry - Stock Market.csv")
data.head()

Symbol                     Name   Price  Change market cap  PE_ratio
0   BRK-B  Berkshire Hathaway Inc.  321.26   -1.38   709.361B       NaN
1       V                Visa Inc.  230.76   -0.25   483.304B     30.81
2     JPM     JPMorgan Chase & Co.  133.89   -2.16   392.325B      9.90
3      MA  Mastercard Incorporated  380.21   -3.18   359.338B     37.85
4  JPM-PD     JPMorgan Chase & Co.   24.57   -0.08   326.427B      1.82

# 2. Cleaning the Data

In [18]:
# Get rid of missing values
data.dropna(inplace=True)
data.head()

Symbol                     Name   Price  Change market cap  PE_ratio
1       V                Visa Inc.  230.76   -0.25   483.304B     30.81
2     JPM     JPMorgan Chase & Co.  133.89   -2.16   392.325B      9.90
3      MA  Mastercard Incorporated  380.21   -3.18   359.338B     37.85
4  JPM-PD     JPMorgan Chase & Co.   24.57   -0.08   326.427B      1.82
5  JPM-PC     JPMorgan Chase & Co.   25.02   -0.09   328.207B      1.86

In [19]:
# Original Data Types
print(data.dtypes)

Symbol         object
Name           object
Price          object
Change        float64
market cap     object
PE_ratio      float64
dtype: object


Converting the object data types to categorical and numeric data types allows me to do further analysis. In the following code chunk, I will convert "Price" and "market cap" to numeric, and "Symbol" and "Name" to categorical.

In [20]:
# Price and market cap must be numeric in order to do analysis

#  Convert "Price" column to numeric using str.replace
data['Price'] = data['Price'].str.replace(',', '').astype(float)

# Convert "market cap" column to float, this removes B and converts to numeric
data['market cap'] = data['market cap'].str.rstrip('B').astype(float)

# Convert market cap to billions
data['market cap'] = data['market cap'] * 1e9  #billion


# Convert the remaining object Data Types to categorical

# This selects the object columns
obj_cols = data.select_dtypes(include=['object'])

# Converts them to categorical data types
data[obj_cols.columns] = obj_cols.astype('category')

# Look at the data types after converting
print(data.dtypes)



Symbol        category
Name          category
Price          float64
Change         float64
market cap     float64
PE_ratio       float64
dtype: object


Next, I check to see whether there are any duplicate rows. If so, they should be removed.

In [21]:
# Original number of rows and columns
print("Original dimensions of the dataset:", data.shape)


Original dimensions of the dataset: (1235, 6)


In [22]:
# Find any duplicate rows
dup_rows = data[data.duplicated()]
print("Duplicate Rows:")
print(dup_rows)

# remove duplicate rows, keeps the first instance
data.drop_duplicates(subset=['Symbol', 'Name'], keep='first', inplace=True)




Duplicate Rows:
      Symbol                                Name    Price  Change  \
28    JPM-PD                JPMorgan Chase & Co.    24.57   -0.08   
30    BAC-PK         Bank of America Corporation    24.38   -0.04   
31    BAC-PL         Bank of America Corporation  1144.76    0.79   
32    BML-PG         Bank of America Corporation    20.02   -0.03   
33    BML-PH         Bank of America Corporation    19.90    0.05   
...      ...                                 ...      ...     ...   
1241     AJG           Arthur J. Gallagher & Co.   216.99    0.18   
1242    BSBR       Banco Santander (Brasil) S.A.     6.10    0.08   
1245  AIG-PA  American International Group, Inc.    24.59    0.02   
1247  MET-PA                       MetLife, Inc.    21.98    0.22   
1248  MET-PE                       MetLife, Inc.    23.79    0.04   

        market cap  PE_ratio  
28    3.264270e+11      1.82  
30    2.445960e+11      6.94  
31    0.000000e+00    325.77  
32    2.703500e+11      5.70  


In [23]:
print("Current dimensions of the dataset:", data.shape)
data.head()

Current dimensions of the dataset: (72, 6)


Symbol                     Name   Price  Change    market cap  PE_ratio
1       V                Visa Inc.  230.76   -0.25  4.833040e+11     30.81
2     JPM     JPMorgan Chase & Co.  133.89   -2.16  3.923250e+11      9.90
3      MA  Mastercard Incorporated  380.21   -3.18  3.593380e+11     37.85
4  JPM-PD     JPMorgan Chase & Co.   24.57   -0.08  3.264270e+11      1.82
5  JPM-PC     JPMorgan Chase & Co.   25.02   -0.09  3.282070e+11      1.86

After removing the duplicate rows, the dataset is now 425 rows x 6 columns.

In order to make the data look more tidy, I will also change the titles of the columns so they appear more consistent.


In [24]:
# Change the titles
data.rename(columns={'market cap': 'Market Cap', 'PE_ratio': 'PE Ratio'}, inplace=True)
data.head()


Symbol                     Name   Price  Change    Market Cap  PE Ratio
1       V                Visa Inc.  230.76   -0.25  4.833040e+11     30.81
2     JPM     JPMorgan Chase & Co.  133.89   -2.16  3.923250e+11      9.90
3      MA  Mastercard Incorporated  380.21   -3.18  3.593380e+11     37.85
4  JPM-PD     JPMorgan Chase & Co.   24.57   -0.08  3.264270e+11      1.82
5  JPM-PC     JPMorgan Chase & Co.   25.02   -0.09  3.282070e+11      1.86

# 3. Exploratory Data Analysis

In [25]:
# 1 Find the number of observations
num_ob = data.shape[0]
print("The total number of observations is", num_ob)

The total number of observations is 72


Since the total observations are 425, this dataset is of a medium size.

In [26]:
# 2 Unique values in each category

# columns
cols = ['Symbol', 'Name', 'Price', 'Change', 'Market Cap', 'PE Ratio']

# for loop to find unique values in each column
for c in cols:
    unique_values = data[c].nunique()
    print("The number of unique values in the", c, "column is", unique_values)



The number of unique values in the Symbol column is 72
The number of unique values in the Name column is 42
The number of unique values in the Price column is 72
The number of unique values in the Change column is 49
The number of unique values in the Market Cap column is 72
The number of unique values in the PE Ratio column is 71


Note how there are 42 names of companies and 72 symbols. This is because some companies have multiple ticker symbols. This is likely due to the different types of stocks the company has available. There are several unique values in the price, change, market cap, and PE ratio columns. This indicates that stocks are traded at many different prices. The unique values in change tells me that there are some stocks with very little daily change, but some stocks with a higher amount of daily change. The market cap unique values tells me that there's a range of market capitilizations across companies. Similarly, there is likely variability present in the PE ratio given its large number of unique values.

In [27]:
# 3 Calculate the number of missing values per variable
missing_per_variable = data.isnull().sum()
print("Missing data per variable:")
print(missing_per_variable)

# Calculate the number of missing values per observation
missing_per_observation = data.isnull().sum(axis=1) # observation
print("Missing data per observation:")
print(missing_per_observation)



Missing data per variable:
Symbol        0
Name          0
Price         0
Change        0
Market Cap    0
PE Ratio      0
dtype: int64
Missing data per observation:
1       0
2       0
3       0
4       0
5       0
       ..
1149    0
1174    0
1199    0
1224    0
1249    0
Length: 72, dtype: int64


Both of these are 0 because I removed null values in the cleaning stage, which is what we want. We will focus on complete data.

In [28]:
# 4 Distributions of continuous variables / Outliers

# Continuous variables in DataFrame
cont_var = ['Price', 'Change', 'Market Cap', 'PE Ratio']  # Update with your continuous variables

# Histograms for each continuous variable
for col in cont_var:
    fig = px.histogram(data, x=col, title=f'Distribution of {col}')
    fig.show()




Histograms Analysis: 

As you can see above, the distribution for Price is right-skewed. This indicates that most of the stocks in this data tend to have lower prices, although there are a few stocks which have very high prices. This is relevant for investors because it can promote investors to invest in multiple different stocks. It would also be relevant when considering a stock is worthwile to invest in.

The distribution for Change is also right-skewed, though most of the companies are near 0. This means that for those stocks, there is not much change on a day-to-day basis. However, there are some stocks with high rates of change. This means that there is a significant change for those stocks on a daily basis. This is relevant for investors because it could provide insight as to whether they are comfortable with small/large changes of their stocks' prices.

The distribution for Market Cap is also right-skewed. This indicates that most of the companies have low market capitalizations, while there are some companies that have very high market capitalizations. Most of the companies seem to have a market capitalization near $100B. These indicate large and well established companies. It would be relevant when considering what stocks to invest in. 

The distribution for PE Ratio is heavily right-skewed. This indicates that most of the companies have lower PE ratios. There are a few stocks that have high PE ratios. This means that the lower PE stocks may show undervaluation. Alternatively, the high PE ratio stocks may show overvaluation. This would be relevant when considering what stocks to invest in, combined with prices.

In [29]:
# 4 Distributions of continuous variables / Outliers Continued
# Boxplot for each continuous variable
for col in cont_var:
    fig = px.box(data, y=col, title=f'Boxplot of {col}')
    fig.show()

Boxplots Analysis:

Price: The median price of the stocks are $62.13. 
The IQR is from $29.12 to $147.97. 
There is one outlier present at a stock price of $1,144.76. This indicates that while most stock prices are near each other, there is one very high stock price. 
I will keep this outlier in the dataset because it could be interesting to visualize in the final dashboard, and may be relevant to potentially invest in.

Change: The median daily change is -0.03. The IQR is from -0.63 to 0.1125. 
There is an outlier of 8.97. 
This indicates that most of the changes in stock prices tend to be relatively small. 
Also, it indicates that on a daily basis, most of the stock prices decrease. 
The outlier means that there is an occurance where the daily change is extremely large. 
This is relevant because it could be the result of something important that happened to the company. 
Hence, I will keep the outlier in the dataset because it could be relevant for investing purposes. 

Market Cap: The median price of the market capitalization is $88.99B. 
The IQR is from $67.01B to $114.93B. There is an outlier of $483.44B. 
This indicates that most of the market capitalization values are of near the median price range. 
However, the outlier could represent a company that is very large, or has performed very well in industry. 
Therefore, it is very relevant to keep this, as this could be important for investing decisions. 

PE Ratio: The median PE Ratio is 14.31. 
The IQR is from 8.12 to 28.18. 
There is an outlier of 325.77. 
This indicates that most of the companies in the dataset have a moderate price to earning ratio. 
The outlier could represent a company with overvaluation. 
This is likely in the case that it is the same outlier with a high price. 
Therefore, it is relevant to keep this outlier, as it could allow the user to make an informed investment decision.

In [30]:
# 5 Describe
data.describe()

Price     Change    Market Cap    PE Ratio
count    72.000000  72.000000  7.200000e+01   72.000000
mean    112.255139  -0.032222  1.222250e+11   21.961806
std     209.010484   1.288593  9.622536e+10   48.235501
min       5.410000  -3.180000  0.000000e+00    0.390000
25%      22.212500  -0.130000  6.037800e+10    4.417500
50%      25.265000  -0.010000  7.960950e+10    8.550000
75%     109.065000   0.097500  1.454595e+11   19.782500
max    1144.760000   8.490000  4.833040e+11  325.770000

Price: The average price is approximately $123.25. 
The prices of the stocks typically range from $5.41 to $1144.76.
The standard deviation is relatively high at $149.85, which indicates significant variability in prices.

Change: The average change is approximately -0.04. 
The daily changes of the stocks typically range from -3.18 to 8.97.
The standard deviation is relatively high at 1.75, which indicates significant variability in changes.

Market Cap: The average market capitalization is approximately $104.80B. 
The market capitalizations of the companies typically range from $0 to $483B.
The standard deviation is relatively high at $65.49B, which indicates significant variability in market capitalizations.

PE Ratio: The average PE ratio is approximately 24.78. 
The PE ratios of the companies typically range from 0.39 to 325.77. 
The standard deviation is relatively high at 30.59, which indicates a significant variability in PE ratios.


In [31]:
# 6 Correlation
num_data = data.select_dtypes(include='number')
corr_mat = num_data.corr()

print(corr_mat)


               Price    Change  Market Cap  PE Ratio
Price       1.000000  0.279955    -0.00848  0.850646
Change      0.279955  1.000000    -0.14216  0.119174
Market Cap -0.008480 -0.142160     1.00000 -0.097220
PE Ratio    0.850646  0.119174    -0.09722  1.000000


Correlations:

Price vs Change: 0.44 is moderately positive which indicates a positive correlation. As the stock price increases, the daily change is also likely to increase. 

Price vs Market Cap: -0.0195 is negative but close to 0, indicating that there is not a particularly strong correlation between the two variables.

Price vs PE Ratio: 0.409 is moderately positive which indicates a positive correlation. As the stock price increases, the PE Ratio is also likely to increase.

Change vs Market Cap: -0.0815 is negative and indicates a potential slight negative correlation. As companies with larger market capitalizations increase, daily price changes may slightly decrease.

Change vs PE Ratio: -0.0105 is negative but close to 0, indicating that there is not a particularly strong correlation between the two variables.

Market Cap vs. PE Ratio: -0.14 is negative and indicates a negative correlation. As companies with larger market capitalizations increase, the PE Ratios decrease, meaning their valuations go down. 

Note that these are correlations. Many of these variables may reveal important information when further visualized. 



It is clear stocks and investments can be risky. Looking at multiple variables, correlations, trends, and patterns are necessary to make well-informed decisions about which stocks to invest in. We can thus utilize this exploratory data analysis to further display interesting and signficant visualizations.

## 4. Data Dictionary

| Column # | Variable    | Description                          | Data Type | Units of Measurement |
|----------|-------------|--------------------------------------|-----------|----------------------|
| 1        | Symbol      | The ticker symbol of the company's stock.                  | Categorical | N/A |
| 2        | Name        | The name of the company.                         | Categorical | N/A |
| 3        | Price       | The most recent trading price of the company's stock.           | Numeric | USD (per share) |
| 4        | Change      | The change in the stock price from the previous trading day's closing price.                | Numeric | USD (per share) |
| 5        | Market Cap  | The market capitalization of the company. This is the total value of all outstanding shares of the company's stock. | Numeric | USD (in billions) |
| 6        | PE Ratio    | The price-to-earnings ratio of the company. This is a measure of the stock's valuation relative to its earnings per share.  | Numeric | N/A |


# 5. List of Potential UI Components:
1. Multi-select dropdown Menu for Name
2. Multi-select dropdown menu for Symbol
3. Slider for Price
4. Grid layout that each has separate functionalities
5. Range slider for Market Capitalization
6. Tooltips on Market Cap and PE Ratio
7. Radio buttons for different filters like Symbol/Name
8. Interactive table of top 5 most recommended stocks with relevant financial info
9. Interactive table of top 5 least recommended stocks with relevant financial info

# 6. List of Potential Data Visualizations
1. Scatter plot of Price vs. Market Capitalization with the option to filter by price range.
2. Histogram of PE Ratio distribution with the option to filter by a range of market capitalization values.
3. Grouped bar chart of Price, Change, and PE Ratio, and allow the user to choose by company name/symbol.
4. Scatter plot of Change vs. PE Ratio and allow the user to filter by a price range.
5. Bubble chart of Price vs. Market Capitalization and the size of the bubble could be based on change.
6. Bubble chart of Market Capitalization vs. PE Ratio and the size of the bubble is based on price.

In [32]:
# Save the new data file to csv
data.to_csv('data.csv', index=False)
